src/libsyntax/parse/parser.rs

```diff
@@ -2902,6 +2904,51 @@ impl<'a> Parser<'a> {
         self.maybe_recover_from_bad_qpath(expr, true)
     }
 
+    fn should_parse_struct_expr(
+        &mut self,
+        lo: Span,
+        path: &ast::Path,
+        attrs: &ThinVec<Attribute>,
+    ) -> Option<PResult<'a, P<Expr>>> {
+        let could_be_struct = self.look_ahead(1, |t| t.is_ident()) && (
+            self.look_ahead(2, |t| *t == token::Colon)
```

Comment:
Type ascription will be mis-parsed (`if x == y { z: TYPE }`).

Reply:
`{ IDENT ,` is the only case where we know it's a struct literal reliably.

Reply:
Additionally, we can cover cases like `{ IDENT: NOT_A_TYPE` by checking that the token following the colon is not `can_begin_type`.
This notably doesn't cover `{ x: y }` though, but covers `{ x: 1 }` or `{ x: -y }`.

Reply:
We can also cover cases like `{ x: y.method() }` with a more elaborated scheme with rollback:
```
save_state()
stmt = parse_stmt() // not speculative enough, can report errors unconditionally
parse_struct = stmt.is_err()
rollback()
if parse_struct { parse_struct() }
```
, but I wouldn't go in this direction personally.

Reply:
I think another alternative would be to have a small pseudoparser that looks for multiple occurrences of `<ident>: <expr>,` until finding a closing `}`, but that is not something I'm inclined to include in this PR.